### ISSUE TITLE
In SQL exporting, what is the order of constraints?

### ISSUE THREAD

**Issue by onaran** [sqlitebrowser/sqlitebrowser#2627]
Hello,
The database imported from sql file does not produce the same order of the table constraints.
I have a table like this
```sql
CREATE TABLE IF NOT EXISTS "CodeLookup" (
	"ID"	INTEGER NOT NULL UNIQUE,
	"Type"	TEXT NOT NULL,
	"Name"	TEXT NOT NULL,
	"VariableName"	TEXT UNIQUE,
	"HexValue"	TEXT,
	"DecValue"	INTEGER NOT NULL,
	"Comment"	TEXT,
	"OldName"	TEXT,
	"CName"	TEXT,
	"CSName"	TEXT,
	"MCUName"	TEXT,
	"XLName"	TEXT,
	UNIQUE("Type","VariableName"),
	PRIMARY KEY("ID" AUTOINCREMENT)
);
```
I do the followings:
1. Import it to a database
2. Export this database to another sql file
3. And the resulting sql file produces the same table, however sometimes in different order of the constraints, i.e. order of UNIQUE and PRIMARY KEY switched like shown below.
```sql
CREATE TABLE IF NOT EXISTS "CodeLookup" (
	"ID"	INTEGER NOT NULL UNIQUE,
	"Type"	TEXT NOT NULL,
	"Name"	TEXT NOT NULL,
	"VariableName"	TEXT UNIQUE,
	"HexValue"	TEXT,
	"DecValue"	INTEGER NOT NULL,
	"Comment"	TEXT,
	"OldName"	TEXT,
	"CName"	TEXT,
	"CSName"	TEXT,
	"MCUName"	TEXT,
	"XLName"	TEXT,
	PRIMARY KEY("ID" AUTOINCREMENT)
	UNIQUE("Type","VariableName"),
);
```
I am interested in the order, since I have put it a git repository. Everytime I follow the procedure above, even if I did not changed another part, the table definition is changed.

In short, how do you determine the order of constraints like unique and primary key? Is that possible to produce the same sql file everytime I export the database?
Thank you so much.

**Comment by MKleusberg** [comment]
Good point. As you could see the order of the constraints is not stable at the moment. I'd love to rewrite the code involved in this and a stable order of the constraints would only be one of the effects of that. But since that requires some more time I have now tried to tweak the existing code to produce a more stable export.

Can you test again using tomorrow's nightly build and see if it already improves the situation for you? You can download that [here](https://nightlies.sqlitebrowser.org/latest/).

**Comment by onaran** [comment]
Thanks for response, I downloaded the one built on March 21. I am going to keep testing for a while and leave a comment here.

**Comment by MKleusberg** [comment]
@onaran Any results yet? :smile:

**Comment by onaran** [comment]
Hi,
I downloaded the nightly build, but I had other issues that I do not remember right now. Therefore did not have a chance to test. 

Downloaded the new version and run a test.
I create a new empty database and instead of using GUI to create a table, I create the table by running the following sql file.
```sql
BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS "CodeLookup" (
	"ID"	INTEGER NOT NULL UNIQUE,
	"Type"	TEXT NOT NULL,
	"Name"	TEXT NOT NULL,
	"VariableName"	TEXT UNIQUE,
	"HexValue"	TEXT,
	"DecValue"	INTEGER NOT NULL,
	"Comment"	TEXT,
	"OldName"	TEXT,
	"CName"	TEXT,
	"CSName"	TEXT,
	"MCUName"	TEXT,
	"XLName"	TEXT,
	UNIQUE("Type","VariableName"),
	PRIMARY KEY("ID" AUTOINCREMENT)
);
COMMIT;
```
Then I immediately export the sql file and the resulting is the same as above. However when I add a new table using the program's create new table button, I got the following.
```sql
CREATE TABLE IF NOT EXISTS "CodeLookup" (
	"ID"	INTEGER NOT NULL UNIQUE,
	"Type"	TEXT NOT NULL,
	"Name"	TEXT NOT NULL,
	"VariableName"	TEXT UNIQUE,
	"HexValue"	TEXT,
	"DecValue"	INTEGER NOT NULL,
	"Comment"	TEXT,
	"OldName"	TEXT,
	"CName"	TEXT,
	"CSName"	TEXT,
	"MCUName"	TEXT,
	"XLName"	TEXT,
	PRIMARY KEY("ID" AUTOINCREMENT),
	UNIQUE("Type","VariableName")
);
CREATE TABLE IF NOT EXISTS "Table" (
	"Field1"	INTEGER
);
```
As you can see from the result, the contraints are swapped.  
I also deleted a field from CodeLookup table I got,
```sql
CREATE TABLE IF NOT EXISTS "Table" (
	"Field1"	INTEGER
);
CREATE TABLE IF NOT EXISTS "CodeLookup" (
	"ID"	INTEGER NOT NULL UNIQUE,
	"Type"	TEXT NOT NULL,
	"Name"	TEXT NOT NULL,
	"HexValue"	TEXT,
	"DecValue"	INTEGER NOT NULL,
	"Comment"	TEXT,
	"OldName"	TEXT,
	"CName"	TEXT,
	"CSName"	TEXT,
	"MCUName"	TEXT,
	"XLName"	TEXT,
	UNIQUE("Type"),
	PRIMARY KEY("ID" AUTOINCREMENT)
);
```
This time not only the constraints are swapped, but also the modified table moved to the end of the file. 
What I wish to have that the tables, views are sorted so that the exported file is not change too much.
Thanks.

**Comment by chrisjlocke** [comment]
> Then I immediately export the sql file

Tables are exported based on how they are represented internally, so its like selecting records from a normal SQLite table - they're going to be in a random order.  Unfortunately there is no way to specify a 'sort order'.  This can cause issues though - sometimes a create table specifies a constraint with another table which doesn't exist yet, so that fails.

**Comment by onaran** [comment]
So there is no way to fix order of the constraints, doing so may create new issues. 
Thanks for the information that you provide.

**Comment by mgrojo** [comment]
@onaran , I've implemented a new option in the Export to SQL dialog to "Keep original CREATE statements", that will keep the order of constraints and any other particularity of the original statements, like comments, quoting or indentation. Could you give it a try with the next [nightly build](https://nightlies.sqlitebrowser.org/latest/) and report any malfunction?

**Comment by onaran** [comment]
Thank you so much, I downloaded the latest nightly build today and keep testing it. Any issues faced will be reported here.
